plugins/quickbooks-online/src/src/Facade/QuickBooksFacade.php

```diff
@@ -376,143 +357,486 @@ function (array $a, array $b) {
-                        }
-
-                        $this->handleErrorResponse($dataService);
                     } catch (\Exception $exception) {
                         $this->logger->error(
                             sprintf(
-                                'Payment ID: %s export failed with error %s.',
-                                $ucrmPayment['id'],
+                                'Payment to apply credit export failed with error %s',
                                 $exception->getMessage()
                             )
                         );
                     }
                 }
             }
 
-            // update last processed payment in the end
-            $pluginData->lastExportedPaymentID = $ucrmPayment['id'];
+            $pluginData->lastExportedCreditID = $ucrmCredit['id'];
             $this->optionsManager->update();
+        } // end foreach $ucrmCredit
+    }
+
+    private function getPaymentCovers($payments, DataService $dataService): array {
+        $lineArray = null;
+        $additionalUnapplied = 0.0;
+        $totalApplied = 0.0;
+        foreach ($payments as $paymentCovers) {
+            if ($paymentCovers['amount'] == 0) continue;
+
+            if ($paymentCovers['refundId']) {
+                $this->logger->notice('Payment has refundId, not yet supported! Will set as unapplied');
+                $additionalUnapplied += $paymentCovers['amount'];
+                continue;
+            }
+            if ($paymentCovers['invoiceId'] == '') {
+                $this->logger->notice('Payment has empty invoiceId! Will set as unapplied');
+                $additionalUnapplied += $paymentCovers['amount'];
+                continue;
+            }
+
+            $invId = $paymentCovers['invoiceId'];
+            $invoices = $this->dataServiceQuery($dataService,"SELECT * FROM INVOICE WHERE DOCNUMBER LIKE '%/$invId'");
+            if (!$invoices)
+                $invoices = $this->dataServiceQuery($dataService,"SELECT * FROM INVOICE WHERE DOCNUMBER = '$invId'");
+
+            if (!$invoices) {
+                $this->logger->warning(sprintf('Unable to find invoiceId %s covered by payment, will set as unapplied', $paymentCovers['invoiceId']));
+                $additionalUnapplied += $paymentCovers['amount'];
+                continue;
+            }
+
+            $lineArray[] = Line::create(
+                [
+                    'Amount' => $paymentCovers['amount'],
+                    'LinkedTxn' => [
+                        'TxnId' => $invoices[0]->Id,
+                        'TxnType' => 'Invoice',
+                    ],
+                ]
+            );
+
+            $totalApplied += $paymentCovers['amount'];
+            $this->logger->debug("Payment applying \${$paymentCovers['amount']} to Invoice {$invoices[0]->DocNumber}");
+        }
+
+        return [$lineArray, $additionalUnapplied, $totalApplied];
+    }
+
+    private $itemCache = [];
```

Comment:
Please put class variables to the beginning of file, don't declare new variables in the middle of methods. If you want to have variables near the methods using them, it's usually a good idea to extract the variables and the method to a new class, it even makes the code more readable.

Reply:
Done